spec-docs/function-types.md

```diff
@@ -0,0 +1,342 @@
+`Function0`, `Function1`, ..., `Function22` and `FunctionN` for 23+ parameters.
+
+## Function trait
+
+There's also an empty trait `kotlin.Function<R>` for cases when e.g. you're storing functions with different/unknown arity
+in a collection to invoke them reflectively somewhere else.
+
+``` kotlin
+package kotlin
+
+trait Function<out R>
+```
+
+It's a physical trait, declared in platform-agnostic built-ins, and present in `kotlin-runtime.jar` for example.
+However its declaration is **empty** and should be empty because every physical JVM function class `Function0`, `Function1`, ...
+inherits from it (and adds `invoke()`), and we don't want to override anything besides `invoke()` when doing it from Java code.
+
+## Functions with 0..22 parameters at runtime
+
+There are 23 function traits in `kotlin.platform.jvm`: `Function0`, `Function1`, ..., `Function22`.
+Here's `Function1` declaration, for example:
+
+``` kotlin
+package kotlin.platform.jvm
+
+trait Function1<in P1, out R> : kotlin.Function<R> {
+    fun invoke(p1: P1): R
+}
+```
+
+These traits are supposed to be inherited from by Java classes when passing lambdas to Kotlin.
+
+Package `kotlin.platform.jvm` is supposed to contain interfaces which help use Kotlin from Java.
+(And not from Kotlin, because normally you would use a function type there,
+most of the time even without mentioning built-in function classes: `(P1, P2, P3) -> R`.)
+
+## Translation of Kotlin lambdas
+
+There's also `FunctionImpl` abstract class at runtime which helps in implementing `arity` and vararg-invocation.
+It inherits from all the physical function classes, unfortunately (more on that later).
+
+``` kotlin
+package kotlin.jvm.internal
+
+abstract class FunctionImpl(override val arity: Int) :
+    Function<Any?>,
+    Function0<Any?>, Function1<Any?, Any?>, ..., ..., Function22<...>,
+    FunctionN   // See the next section on FunctionN
+{
+    override fun invoke(): Any? {
+        // Default implementations of all "invoke"s invoke "invokeVararg"
+        // This is needed for KFunctionImpl (see below)
+        assert(arity == 0)
+        return invokeVararg()
+    }
+    
+    override fun invoke(p1: Any?): Any? {
+        assert(arity == 1)
+        return invokeVararg(p1)
+    }
+    
+    ...
+    override fun invoke(p1: Any?, ..., p22: Any?) { ... }
+    
+    override fun invokeVararg(vararg p: Any?): Any? = throw UnsupportedOperationException()
+    
+    override fun toString() = ... // Some calculation involving generic runtime signatures
+}
+```
+
+> TODO: sadly, this class needs to be implemented in Java because supertypes need to be **raw** classes
+> for reflection to pick up correct generic signatures for inheritors
+
+Each lambda is compiled to an anonymous class which inherits from `FunctionImpl` and implements the corresponding `invoke`:
+
+``` kotlin
+{ (s: String): Int -> s.length }
+
+// is translated to
+
+object : FunctionImpl(2), Function2<String, Int> {
```

Comment:
Doesn't `Function2` have 3 type parameters (2 for parameter types and 1 for return type)?